drivers/flash/flash_qspi_renesas_ra.c

```diff
@@ -0,0 +1,501 @@
+};
+
+static const struct flash_parameters qspi_flash_ra_config_para = {
+	.write_block_size = QSPI_WRITE_BLK_SZ,
+	.erase_value = 0xff,
+};
+
+static const qspi_extended_cfg_t g_qspi_extended_cfg = {
+	.min_qssl_deselect_cycles = QSPI_QSSL_MIN_HIGH_LEVEL_8_QSPCLK,
+	.qspclk_div = QSPI_QSPCLK_DIV_2,
+};
+
+static struct qspi_flash_ra_data qspi_flash_data = {
+	.qspi_cfg = {
+		.spi_protocol = SPI_FLASH_PROTOCOL_EXTENDED_SPI,
+		.read_mode = SPI_FLASH_READ_MODE_FAST_READ_QUAD_IO,
+		.address_bytes = SPI_FLASH_ADDRESS_BYTES_3,
+		.dummy_clocks = SPI_FLASH_DUMMY_CLOCKS_DEFAULT,
+		.page_program_address_lines = SPI_FLASH_DATA_LINES_1,
+		.page_size_bytes = PAGE_SIZE_BYTE,
+		.page_program_command = QSPI_CMD_PAGE_PROGRAM,
+		.write_enable_command = QSPI_CMD_WRITE_ENABLE,
+		.status_command = QSPI_CMD_READ_STATUS,
+		.write_status_bit = WRITE_STATUS_BIT,
+		.xip_enter_command = QSPI_CMD_XIP_ENTER,
+		.xip_exit_command = QSPI_CMD_XIP_EXIT,
+		.p_erase_command_list = &g_qspi_erase_command_list[0],
+		.erase_command_list_length = ERASE_COMMAND_LENGTH(g_qspi_erase_command_list),
+		.p_extend = &g_qspi_extended_cfg,
+	}};
+
+static void acquire_device(const struct device *dev)
+{
+	struct qspi_flash_ra_data *dev_data = dev->data;
+
+	k_sem_take(&dev_data->sem, K_FOREVER);
+}
+
+static void release_device(const struct device *dev)
+{
+	struct qspi_flash_ra_data *dev_data = dev->data;
+
+	k_sem_give(&dev_data->sem);
+}
+static int get_flash_status(const struct device *dev)
+{
+	struct qspi_flash_ra_data *qspi_data = dev->data;
+	spi_flash_status_t status = {.write_in_progress = true};
+	int32_t time_out = (INT32_MAX);
+	int err;
+
+	do {
+		err = R_QSPI_StatusGet(&qspi_data->qspi_ctrl, &status);
+		if (err != FSP_SUCCESS) {
+			LOG_ERR("qspi: status get failed");
+			return -EIO;
+		}
+		--time_out;
+		if (RESET_VALUE >= time_out) {
+			return -EIO;
+		}
+	} while (false != status.write_in_progress);
+
+	return 0;
+}
+
+#if defined(CONFIG_FLASH_EX_OP_ENABLED)
+static int qspi_flash_ra_ex_op(const struct device *dev, uint16_t code, const uintptr_t in,
+			       void *out)
+{
+	int err = 0;
+	uint8_t cmd;
+	struct qspi_flash_ra_data *qspi_data = dev->data;
+
+	ARG_UNUSED(in);
+	ARG_UNUSED(out);
+	acquire_device(dev);
+
+	switch (code) {
+	case QSPI_FLASH_EX_OP_EXIT_QPI:
+		if (SPI_FLASH_PROTOCOL_QPI != qspi_data->qspi_cfg.spi_protocol) {
+			err = 0;
+			break;
+		}
+		cmd = QSPI_CMD_EXIT_QPI_MODE;
+		err = R_QSPI_DirectWrite(&qspi_data->qspi_ctrl, &cmd, ONE_BYTE, false);
+		if (err != FSP_SUCCESS) {
+			LOG_ERR("qspi: direct write failed");
+			err = -EIO;
+		}
+		break;
+
+	case QSPI_FLASH_EX_OP_RESET:
+		cmd = QSPI_CMD_RESET_EN;
+		err = R_QSPI_DirectWrite(&qspi_data->qspi_ctrl, &cmd, ONE_BYTE, false);
+		if (err == FSP_SUCCESS) {
+			cmd = QSPI_CMD_RESET_MEM;
+			err = R_QSPI_DirectWrite(&qspi_data->qspi_ctrl, &cmd, ONE_BYTE, false);
+			if (err != FSP_SUCCESS) {
+				LOG_ERR("qspi: direct write failed");
+				err = -EIO;
+			}
```

Comment:
You lack here userspace processing, is that desired?

Reply:
Yes, it is designed without userspace processing.
It is supported to reset QPI mode and memory.